tests/lib/openstack_query/test_query_parser.py

```diff
@@ -0,0 +1,415 @@
+    def test_run_parser_with_sort_and_group(self, mock_run_group_by, mock_run_sort):
+        """
+        Tests that run_parser functions expectedly - when giving both group_by and sort_by
+        Should call run_sort method and then run_group_by on that output, then return
+        """
+        self.instance._sort_by = (MockProperties.PROP_1, False)
+        self.instance._group_by = MockProperties.PROP_1
+        self.instance._group_mappings = {"group1": "some-mapping_func"}
+        mock_obj_list = ["obj1", "obj2", "obj3"]
+
+        mock_run_group_by.return_value = "group-out"
+        mock_run_sort.return_value = "sort-out"
+
+        res = self.instance.run_parser(mock_obj_list)
+        self.assertEqual(res, "group-out")
+        mock_run_sort.assert_called_once_with(mock_obj_list, self.instance._sort_by)
+        mock_run_group_by.assert_called_once_with(
+            "sort-out", self.instance._group_mappings
+        )
+
+    @parameterized.expand(
+        [
+            (
+                "string key ascending",
+                {"arg1": False},
+                [
+                    {"arg1": "a", "arg2": 2},
+                    {"arg1": "b", "arg2": 1},
+                    {"arg1": "c", "arg2": 4},
+                    {"arg1": "d", "arg2": 3},
+                ],
+            ),
+            (
+                "string key descending",
+                {"arg1": True},
+                [
+                    {"arg1": "d", "arg2": 3},
+                    {"arg1": "c", "arg2": 4},
+                    {"arg1": "b", "arg2": 1},
+                    {"arg1": "a", "arg2": 2},
+                ],
+            ),
+            (
+                "integer key ascending",
+                {"arg2": False},
+                [
+                    {"arg1": "b", "arg2": 1},
+                    {"arg1": "a", "arg2": 2},
+                    {"arg1": "d", "arg2": 3},
+                    {"arg1": "c", "arg2": 4},
+                ],
+            ),
+            (
+                "integer key descending",
+                {"arg2": True},
+                [
+                    {"arg1": "c", "arg2": 4},
+                    {"arg1": "d", "arg2": 3},
+                    {"arg1": "a", "arg2": 2},
+                    {"arg1": "b", "arg2": 1},
+                ],
+            ),
+        ]
+    )
+    def test_run_sort_with_one_key(self, _, mock_sort_by_specs, expected_res):
+        """
+        Tests that run_sort functions expectedly - with one sorting key
+        Should call run_sort method which should get the appropriate sorting
+        key lambda function and sort dict accordingly
+        """
+
+        mock_obj_list = [
+            {"arg1": "a", "arg2": 2},
+            {"arg1": "c", "arg2": 4},
+            {"arg1": "d", "arg2": 3},
+            {"arg1": "b", "arg2": 1},
+        ]
+
+        res = self.instance._run_sort(mock_obj_list, sort_by_specs=mock_sort_by_specs)
```

Comment:
`_run_sort` implies an inner class for output parsing, either we should add a `sort_results` to the query_parser or split this out

Reply:
we can split this two parts - `QueryParserBuilder` and `QueryParserRunner`
but that will mean we should probably also split `QueryOutput` into `QueryOutputBuilder` and `QueryOutputRunner`

I'll create an issue for this